src/native/eventpipe/ep-session.c

```diff
@@ -523,6 +620,246 @@ ep_session_write_all_buffers_to_file (EventPipeSession *session, bool *events_wr
 	return !ep_file_has_errors (session->file);
 }
 
+static
+void
+session_disable_user_events (EventPipeSession *session)
+{
+	EP_ASSERT (session != NULL);
+	ep_return_void_if_nok (session->session_type == EP_SESSION_TYPE_USEREVENTS && session->user_events_data_fd > 0);
+
+	ep_requires_lock_held ();
+
+	EventPipeSessionProviderList *providers = session->providers;
+	EP_ASSERT (providers != NULL);
+	DN_LIST_FOREACH_BEGIN (EventPipeSessionProvider *, session_provider, ep_session_provider_list_get_providers (providers)) {
+		ep_session_provider_unregister_tracepoints (session_provider, session->user_events_data_fd);
+	} DN_LIST_FOREACH_END;
+
+#if HAVE_UNISTD_H
+	close (session->user_events_data_fd);
+#endif // HAVE_UNISTD_H
+	session->user_events_data_fd = 0;
+}
+
+static
+bool
+is_guid_empty(const uint8_t *guid)
+{
+	if (guid == NULL)
+		return true;
+
+	for (size_t i = 0; i < EP_ACTIVITY_ID_SIZE; ++i) {
+		if (guid[i] != 0)
+			return false;
+	}
+
+	return true;
+}
+
+/*
+ *  construct_extension_activity_ids_buffer
+ *
+ *  Encodes non-empty activity ID and related activity ID.
+ */
+static
+uint16_t
+construct_extension_activity_ids_buffer (
+	uint8_t *extension,
+	size_t extension_size,
+	const uint8_t *activity_id,
+	const uint8_t *related_activity_id)
+{
+	uint16_t offset = 0;
+
+	memset (extension, 0, extension_size);
```

Comment:
Should we at least assert that the size is large enough?

Reply:
Added asserts that extension is not NULL and that extension_size is large enough to hold the two activity_ids